cabal-install/Distribution/Solver/Modular.hs

```diff
@@ -116,33 +123,56 @@ solve' :: SolverConfig
        -> Set PN
        -> Progress String String (Assignment, RevDepMap)
 solve' sc cinfo idx pkgConfigDB pprefs gcs pns =
-    foldProgress Step (uncurry createErrorMsg) Done (runSolver printFullLog sc)
+    toProgress $ retry (runSolver printFullLog sc)
+                       (createErrorMsg (solverVerbosity sc) (maxBackjumps sc))
   where
     runSolver :: Bool -> SolverConfig
-              -> Progress String (SolverFailure, String) (Assignment, RevDepMap)
+              -> RetryLog String SolverFailure (Assignment, RevDepMap)
     runSolver keepLog sc' =
-        logToProgress keepLog (solverVerbosity sc') (maxBackjumps sc') $
+        displayLogMessages keepLog $
         solve sc' cinfo idx pkgConfigDB pprefs gcs pns
 
-    createErrorMsg :: SolverFailure -> String
-                   -> Progress String String (Assignment, RevDepMap)
-    createErrorMsg (ExhaustiveSearch cs _) msg =
-        Fail $ rerunSolverForErrorMsg cs ++ msg
-    createErrorMsg BackjumpLimitReached    msg =
-        Step ("Backjump limit reached. Rerunning dependency solver to generate "
+    createErrorMsg :: Verbosity
+                   -> Maybe Int
+                   -> SolverFailure
+                   -> RetryLog String String (Assignment, RevDepMap)
+    createErrorMsg verbosity mbj failure@(ExhaustiveSearch cs cm) =
+      if asBool $ minimizeConflictSet sc
+      then continueWith ("Found no solution after exhaustively searching the "
+                          ++ "dependency tree. Rerunning the dependency solver "
+                          ++ "to minimize the conflict set, {"
+                          ++ showConflictSet cs ++ "}.") $
+           retry (tryToMinimizeConflictSet (runSolver printFullLog) sc cs cm) $
+               \case
+                  ExhaustiveSearch cs' cm' ->
+                      fromProgress $ Fail $
+                          rerunSolverForErrorMsg cs'
+                       ++ finalErrorMsg verbosity mbj (ExhaustiveSearch cs' cm')
+                  BackjumpLimitReached ->
+                      fromProgress $ Fail $
+                          "Reached backjump limit while trying to minimize the "
+                       ++ "conflict set to create a better error message. "
+                       ++ "Original error message:\n"
+                       ++ rerunSolverForErrorMsg cs
+                       ++ finalErrorMsg verbosity mbj failure
+      else fromProgress $ Fail $
```

Comment:
I think that, if the conflict set minimise flag isn't set, we should include a hint in the error that turning it on might help.

Reply:
I didn't add a hint initially because I was worried that the flag wouldn't help in many cases.  I added it for the case where the conflict set has more than three variables, where it has a higher chance of removing variables.